Source/processcontainers/implementations/CRunImplementation/CRunImplementation.cpp

```diff
@@ -0,0 +1,460 @@
+#include "JSON.h"
+#include <thread>
+
+namespace WPEFramework
+{    
+namespace ProcessContainers
+{
+    // NetworkInterfaceIterator
+    // ----------------------------------
+    CRunNetworkInterfaceIterator::CRunNetworkInterfaceIterator()
+        : NetworkInterfaceIterator()
+    {
+        TRACE_L1("CRunNetworkInterfaceIterator::CRunNetworkInterfaceIterator() not implemented")
+    }
+
+    CRunNetworkInterfaceIterator::~CRunNetworkInterfaceIterator()
+    {
+        TRACE_L1("CRunNetworkInterfaceIterator::~CRunNetworkInterfaceIterator() not implemented")
+    }
+
+    std::string CRunNetworkInterfaceIterator::Name() const 
+    {
+        TRACE_L1("CRunNetworkInterfaceIterator::Name() not implemented")
+
+        return "";
+    }
+
+    uint32_t CRunNetworkInterfaceIterator::NumIPs() const 
+    {
+        TRACE_L1("CRunNetworkInterfaceIterator::NumIPs() not implemented")
+
+        return 0;
+    }
+
+    std::string CRunNetworkInterfaceIterator::IP(uint32_t id) const 
+    {
+        TRACE_L1("CRunNetworkInterfaceIterator::IP() not implemented")
+
+        return "";
+    }
+
+    // Container administrator
+    // ----------------------------------
+    IContainerAdministrator& ProcessContainers::IContainerAdministrator::Instance()
+    {
+        static CRunContainerAdministrator& runCContainerAdministrator = Core::SingletonType<CRunContainerAdministrator>::Instance();
+
+        return runCContainerAdministrator;
+    }
+
+    IContainer* CRunContainerAdministrator::Container(const string& id, IStringIterator& searchpaths,  const string& logpath, const string& configuration) 
+    {
+        while (searchpaths.Next()) {
+            auto path = searchpaths.Current();
+
+            Core::File configFile(path + "/Container/config.json");
+
+            if (configFile.Exists()) {
+                CRunContainer* container = new CRunContainer(id, path + "/Container", logpath);
+                _containers.push_back(container);
+                AddRef();
+
+                return container;
+            }
+        }
+
+        return nullptr;
+    }
+
+    CRunContainerAdministrator::CRunContainerAdministrator()
+        : _refCount(1)
+    {
+
+    }
+
+    CRunContainerAdministrator::~CRunContainerAdministrator()
+    {
+        if (_containers.size() > 0) {
+            TRACE_L1("There are still active containers when shutting down administrator!");
+            
+            while (_containers.size() > 0) {
+                _containers.back()->Release();
+                _containers.pop_back();
+            }
+        }
+    }
+
+    void CRunContainerAdministrator::Logging(const string& logPath, const string& loggingOptions)
+    {
+        // Only container-scope logging
+    }
+
+    CRunContainerAdministrator::ContainerIterator CRunContainerAdministrator::Containers()
+    {
+        return ContainerIterator(_containers);
+    }
+
+    void CRunContainerAdministrator::AddRef() const
+    {
+        _refCount++;
```

Comment:
atomic or InterLockedIncrement/Decrement (prefer the latter)
Interesting, I do not think the reference count has any meaning as this is just singleton :-)